### DISCUSSION TITLE
Dictionary/HashMap

### DISCUSSION THREAD

**Discussion by unbounded** [uiua-lang/uiua#297] [Language Development]
Happy Advent of Code! :slightly_smiling_face: 

Just curious if you foresee this language getting a dictionary / hash map type for O(1) lookups?

I tried implementing a static hash map with string keys, but I'm thinking it could be done better with language support for e.g. hashing.

```
# djb2 (+ modulo)
Hash ← ∧(◿2147483647 +×33:-@\0) :5381 °□

---
⍤⊃⋅∘≍ ◿2147483647 210676686969 Hash □ "Hello"
---

# MakeTable <Keys> <Values>
MakeTable ← [⊃(↯)(↯;:) ⊃⊃(¤⧻)(⊕□)(⊕□;:) ◿⧻. ≡Hash .]

# Lookup <Table> <Key>
Lookup ← ⊐⊏⊗ ⊃⋅⋅∘(⊙∘) ⊃(°□⊡⊙(⊡0))(⊡⊙(⊡1)) ◿⧻⊏0,Hash,

---
TT ← MakeTable {"hello" "world" "how" "are" "you"} [1 2 3 4 5]

⍤⊃⋅∘≍ 1 Lookup TT □"hello"
⍤⊃⋅∘≍ 5 Lookup TT □"you"
⍤⊃⋅∘≍ 0 ⬚0Lookup TT □"nada"
⍤⊃⋅∘≍ 0 ⍣Lookup (0;;;) TT □"nada"

MT ← MakeTable ≡(□$"Number _") . ⇡ 1000
⍤⊃⋅∘≍ 99 Lookup MT □"Number 99"

LT ← MakeTable ≡(□$"Number _") . ⇡ 10000
⍤⊃⋅∘≍ 99 Lookup LT □"Number 99"
---
```

**Comment by kaikalii** [comment]
I do not plan on adding hash maps.
It is an array language. Having the only data structure be an array is very powerful.
Some built-in functions use hash maps internally.
These include `classify` and `group`. The idea is to use these in a clever way if you want fast lookups.

**Comment by kaikalii** [comment]
There are now experimental functions that manipulate arrays as if they are hashmaps.
See the [`map`](https://www.uiua.org/docs/map) function.
